emulator/src/core/emulator/instructions/memory.rs

```diff
@@ -0,0 +1,203 @@
+use super::super::{Emulator, EResult};
+
+impl Emulator {
+    pub fn stax(&mut self, register: &str) -> EResult<()> {
+        self.ram[self.reg[register]] = self.reg['a'];
+        Ok(())
+    }
+    
+    pub fn ldax(&mut self, register: &str) -> EResult<()> {
+        self.reg['a'] = self.ram[self.reg[register]];
+        Ok(())
+    }
+    
+    pub fn push(&mut self, val: u16) -> EResult<()> {
+        if self.sp < 2 {
+            return Err("PUSH: No more stack space");
+        }
+        self.sp -= 1;
+        self.ram[self.sp] = (val >> 8) as u8;
+        self.sp -= 1;
+        self.ram[self.sp] = val as u8;
+        Ok(())
+    }
+
+    pub fn push_reg(&mut self, reg: &str) -> EResult<()> {
+        self.push(self.reg[reg])
+    }
+
+    pub fn pop(&mut self) -> EResult<u16> {
+        if self.sp + 2 > self.ram.size() as u16 {
+            return Err("POP: No return address on the stack");
+        }
+        let low = self.ram[self.sp] as u16;
+        self.sp += 1;
+        let high = self.ram[self.sp] as u16;
+        self.sp += 1;
+        Ok((high << 8) | low)
+    }
+    
+    pub fn pop_reg(&mut self, reg: &str) -> EResult<()> {
+        self.reg[reg] = self.pop()?;
+        Ok(())
+    }
+    
+    pub fn shld(&mut self, address: u16) -> EResult<()> {
+        self.ram[address] = self.reg['l'];
+        self.ram[address+1] = self.reg['h'];
+        Ok(())
+    }
+    
+    pub fn lhld(&mut self, address: u16) -> EResult<()> {
+        self.reg['l'] = self.ram[address];
+        self.reg['h'] = self.ram[address+1];
+        Ok(())
+    }
+    
+    pub fn sta(&mut self, address: u16) -> EResult<()> {
+        self.ram[address] = self.reg['a'];
+        Ok(())
+    }
+    
+    pub fn lda(&mut self, address: u16) -> EResult<()> {
+        self.reg['a'] = self.ram[address];
+        Ok(())
+    }
+    
+    pub fn xthl(&mut self) -> EResult<()> {
+        let tempL = self.reg['l'];
+        let tempH = self.reg['h'];
+        self.reg['l'] = self.ram[self.sp];
+        self.reg['h'] = self.ram[self.sp+1];
+        self.ram[self.sp] = tempL;
+        self.ram[self.sp+1] = tempH;
+        Ok(())
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn stax() {
+        let mut e = Emulator::new();
```

Comment:
In my opinion variable names should be longer than a single character (e.g. `emu`). However this is of personal nature, don't let it stop you from merging.

Reply:
Agree, just copied it everywhere